rfcs/47-better-error-handing.md

```diff
@@ -0,0 +1,209 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2020-05-03
+---
+
+# AOS-47: Better Error Handling
+
+## Background
+
+### AOS-11
+
+`11-error-handling` proposed that in the top level, [go-storage] should return custom error `struct`s carrying contextual errors. Specifically, the `struct`s will have the following definitions:
+
+```go
+type SomeError struct {
+	Op  string
+	Err error
+
+	ContextA string
+	ContextB structB
+	...
+}
+```
+
+The `Err` field wraps the cause of the error. 
+- For unexpected error, it will be wrapped directly. 
+- For expected error, the related error should be used directly.
+
+### Current Practice
+
+Currently, the error handling mechanism in [go-storage] can be summarized as below:
+
+- top-level error `struct`s (returned by public APIs): `InitError`, `ServiceError` and `StorageError`
+	- defined in [go-storage/services/error.go]
+	- have `Op`, `Err` and context fields
+	- returned in `defer` in `go-service-*`s' public APIs
+		- Public APIs in `generated.go` (generated by `service.tmpl`) call `s.formatError` in `defer`. So the implementor should implement `formatError` method for `Storage` (and `Service` maybe) where `StorageError` (or `ServiceError`) is returned.
+		- For other handwritten public APIs like `New`, the implementor should similarly return a top-level error like `InitError`.
+	- They are orthogonal (won't wrap each other).
+
+- `Err`s wrapped are either:
+  - unexpected errors: original SDK errors: `*googleapi.Error`, `*qserror.QingStorError`, ...
+  - expected errors, either:
+    - ad-hoc string errors: `errors.New()` or `fmt.Errorf()` without `%w` verb
+    - sentinel errors wrapped by `fmt.Errorf` with `%w`: 
+    	- defined in either [go-storage/services/error.go] or `go-service-*/error.go`
+    	- defined as `var ErrSomethingWrong = errors.New("what happened")`
+    	- returned in `fmt.Errorf("%w: %v", services.ErrObjectNotExist, err)`, carrying error messages of original SDK errors
+    - another layer of error `struct`s carrying contextual information: `MetadataUnrecognizedError`, `PairUnsupportedError`, `PairRequiredError`
+    	- defined in [go-storage/services/error.go]
+    	- have `Err` and context fields, don't have `Op`
+    	- returned by constructor methods, where `Err` is set to a sentinel error, which can be viewed as the error `struct`s' classification
+
+![](https://user-images.githubusercontent.com/37948597/116873067-9ce86c80-ac49-11eb-9409-91dcc84da340.png)
```

Comment:
Please upload images to `rfcs/44`, so that we can link to image like `![](./44/old.png)`